src/transaction/log_append.cpp

```diff
@@ -1569,6 +1571,157 @@ prior_is_tde_encrypted (const log_prior_node *node)
   return node->tde_encrypted;
 }
 
+static size_t
+prior_node_serialized_size (const log_prior_node &node)
+{
+  size_t total_size = 0;
+
+  // fixed data
+  total_size += sizeof (node.log_header)
+		+ sizeof (node.start_lsa)
+		+ sizeof (node.tde_encrypted)
+		+ sizeof (node.data_header_length)
+		+ sizeof (node.ulength)
+		+ sizeof (node.rlength);
+
+  // variable data
+  total_size += static_cast<size_t> (node.data_header_length);
+  total_size += static_cast<size_t> (node.ulength);
+  total_size += static_cast<size_t> (node.rlength);
+
+  return total_size;
+}
+
+static size_t
+prior_list_serialized_size (const log_prior_node *head)
+{
+  size_t total_size = 0;
+  for (const log_prior_node *nodep = head; nodep != nullptr; nodep = nodep->next)
+    {
+      total_size += prior_node_serialized_size (*nodep);
+    }
+  return total_size;
+}
+
+static void
+prior_node_serialize (const log_prior_node &node, std::string &serialized)
+{
+#define copy_mem(mem, memsz) if ((memsz) > 0) serialized.append (mem, memsz)
+#define copy(val) copy_mem (reinterpret_cast<const char *> (&(val)), sizeof (val))
+
+  copy (node.log_header);
+  copy (node.start_lsa);
+  copy (node.tde_encrypted);
+  copy (node.data_header_length);
+  copy (node.ulength);
+  copy (node.rlength);
+
+  copy_mem (node.data_header, node.data_header_length);
+  copy_mem (node.udata, node.ulength);
+  copy_mem (node.rdata, node.rlength);
+
+#undef copy_mem
+#undef copy
+}
+
+static const char *
+prior_node_deserialize (const char *ptr, log_prior_node &node)
+{
+#define copy_mem(mem, memsz) std::memcpy (mem, ptr, memsz); ptr += (memsz)
+#define copy(val) copy_mem (reinterpret_cast<char *> (&(val)), sizeof (val))
+#define alloc_and_copy_mem(mem, memsz)	\
+  if ((memsz) > 0)			\
+    {					\
+      (mem) = (char *) malloc (memsz);	\
+      assert ((mem) != nullptr);	\
+      copy_mem (mem, memsz);		\
+    }					\
+  else					\
+    {					\
+      mem = nullptr;			\
+    }
+
+  copy (node.log_header);
+  copy (node.start_lsa);
+  copy (node.tde_encrypted);
+  copy (node.data_header_length);
+  copy (node.ulength);
+  copy (node.rlength);
+
+  alloc_and_copy_mem (node.data_header, node.data_header_length);
+  alloc_and_copy_mem (node.udata, node.ulength);
+  alloc_and_copy_mem (node.rdata, node.rlength);
+
+  node.next = nullptr;
+
+#undef copy_mem
+#undef copy
+#undef alloc_and_copy_mem
+
+  return ptr;
+}
+
+std::string
+prior_list_serialize (const log_prior_node *head)
+{
+  if (head == nullptr)
+    {
+      return std::string ();
+    }
+
+  // calculate size
+  size_t size = prior_list_serialized_size (head);
+
+  std::string serialized;
+  serialized.reserve (size);
+
+  for (const log_prior_node *nodep = head; nodep != nullptr; nodep = nodep->next)
+    {
+      prior_node_serialize (*nodep, serialized);
+    }
+
+  assert (serialized.size () == size);
+
+  return std::move (serialized);
```

Comment:
I've seen this in many places.
I'm not so knowledgeable into this either but maybe we should investigate how to do it correctly - https://stackoverflow.com/questions/14856344/when-should-stdmove-be-used-on-a-function-return-value - https://duckduckgo.com/?q=c%2B%2B+%22return+std%3A%3Amove%22&t=canonical&ia=web

Reply:
You're right. I removed the std::move.